### PR TITLE
AP_Baro: add driver for the SPL06-001

### PR DESCRIPTION
**Pull request by shellixyz**
The SPL06-001 is a high resolution low noise barometer pin-to-pin compatible with the BMP280.

I have been flight testing this baro and driver it seems to work well.

Here is the datasheet: [SPL06-001_datasheet.pdf](https://github.com/ArduPilot/ardupilot/files/3834844/SPL06-001_datasheet.pdf)

### CODE REVIEW

File: libraries/AP_Baro/AP_Baro.cpp [hunk 1]
```diff
@@ -160,8 +161,8 @@ const AP_Param::GroupInfo AP_Baro::var_info[] = {
     // @Param: PROBE_EXT
     // @DisplayName: External barometers to probe
     // @Description: This sets which types of external i2c barometer to look for. It is a bitmask of barometer types. The I2C buses to probe is based on GND_EXT_BUS. If GND_EXT_BUS is -1 then it will probe all external buses, otherwise it will probe just the bus number given in GND_EXT_BUS.
-    // @Bitmask: 0:BMP085,1:BMP280,2:MS5611,3:MS5607,4:MS5637,5:FBM320,6:DPS280,7:LPS25H,8:Keller,10:MS5837,11:BMP388
-    // @Values: 1:BMP085,2:BMP280,4:MS5611,8:MS5607,16:MS5637,32:FBM320,64:DPS280,128:LPS25H,256:Keller,512:MS5837,1024:BMP388
+    // @Bitmask: 0:BMP085,1:BMP280,2:MS5611,3:MS5607,4:MS5637,5:FBM320,6:DPS280,7:LPS25H,8:Keller,9:MS5837,10:BMP388,11:SPL06
```
[the remark this answers is not in the document]
Comment (shellixyz) [reply libraries/AP_Baro/AP_Baro.cpp]: Ok. I made a new PR for that: #12830

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        case 64: return 1040384;
+        case 128: return 2088960;
+        default: return -1; // invalid
+    }
+}
+
+// acumulate a new sensor reading
+void AP_Baro_SPL06::_timer(void)
+{
+    static int8_t counter = -1;
```
[the remark this answers is not in the document]
Comment (shellixyz) [reply libraries/AP_Baro/AP_Baro_SPL06.cpp]: Good catch. I didn't think about the possibility of more than one instance. Fixed

**Comment by bys1123** [comment]
Wow, glad to see this driver in official repo.

**Comment by tridge** [comment]
rebased ...

**Comment by bys1123** [comment]
Do you consider add SPL06-007 later? The latest version has lower temperature drift. But it has some different on register.
https://datasheet.lcsc.com/szlcsc/1907081118_Goertek-SPL06-007_C233787.pdf

File: libraries/AP_Baro/AP_Baro_SPL06.cpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    _c30 = ((uint16_t)buf[16] << 8) | (uint16_t)buf[17];
+
+    // setup temperature and pressure measurements
+    _dev->setup_checked_registers(3, 20);
+
+    _dev->write_register(SPL06_REG_TEMPERATURE_CFG, SPL06_TEMP_USE_EXT_SENSOR | SPL06_OVERSAMPLING_TO_REG_VALUE(SPL06_TEMPERATURE_OVERSAMPLING), true);
+    _dev->write_register(SPL06_REG_PRESSURE_CFG, SPL06_OVERSAMPLING_TO_REG_VALUE(SPL06_PRESSURE_OVERSAMPLING), true);
+
+    uint8_t int_and_fifo_reg_value = 0;
+    if (SPL06_TEMPERATURE_OVERSAMPLING > 8) {
```
Comment (lucasdemarchi) [inline libraries/AP_Baro/AP_Baro_SPL06.cpp]: isn't this dead code?

**Comment by shellixyz** [comment]
@lucasdemarchi Flight tested yes

File: libraries/AP_Baro/AP_Baro_SPL06.cpp (hunk 3 above)
Comment (shellixyz) [reply libraries/AP_Baro/AP_Baro_SPL06.cpp]: It is dead code in this case but won't be if for any reason the oversampling value is raised

**Comment by shellixyz** [comment]
> Do you consider add SPL06-007 later? The latest version has lower temperature drift. But it has some different on register.
> https://datasheet.lcsc.com/szlcsc/1907081118_Goertek-SPL06-007_C233787.pdf

I don't have any, only 001, but why not :)

**Comment by lucasdemarchi** [comment]
@shellixyz thanks

**Comment by bys1123** [comment]
We wrote this general usage driver in C for SPL06-007: https://github.com/DimianZhan/spl06
FYI @shellixyz

**Pull request merged**